D-Freezer/freezer.md

```diff
@@ -0,0 +1,70 @@
+# Implementing a scheduling class
+
+## Introduction
+One of the reasons students struggle with this assignment boils down to a lack of understanding of how the scheduler works. In this guide, I hope to provide you with a clear understanding of how the Linux scheduler pieces fit together. I hope to paint a picture that you can use to help you implement the freezer scheduler.
+
+## The `task_struct`
+
+Recall from HW1 that in Linux, every process is defined by its `struct task_struct`. When you have multiple tasks forked off a common parent, they are linked together in a doubly linked-list `struct list_head siblings` embedded within the `task_struct`. For example, if you had four processes running on your system, each forked off one parent, it will look something like this (the parent is not shown): 
+
+<div align='center'>
+    <img src='./task_struct.png'/><br/>
+</div>
+
+However, at this stage, none of these processes are actually running on a CPU. In order to get them onto a CPU, I need to introduce you to the `struct rq`. 
+
+## The `struct_rq`
+
+The `struct rq` is a per-cpu run queue data structure. I like to think of it as the virtual CPU. It contains a lot of information (must of which goes way over my head), but it also includes the list of tasks that will (eventually) run on that CPU. 
+So how does that work? 
+A naive implementation would be to embed a `struct list_head runqueue_head` (for example) into the `struct rq`, and embed a `struct list_head node` into every `task_struct`. 
+<div align='center'>
+    <img src='./naive.png'/><br/>
+    This is a BAD implementation. 
+</div>
+
+The main problem with this implementation is that it does not extend well. At this point, you know Linux has more than one scheduling class. Linux comes built with a Deadline class, a Real-time class, and the primary CFS. Having a `list_head` embedded directly into the `struct rq` for each scheduling class is not feasible. The solution is to create a new structure containing the `list_head` and any bookkeeping variables into a separate structure. Then, we can include just the wrapper structure in the `struct rq`. Linux includes these structures in `linux/kernel/sched/sched.h`. 
+
+By convention, Linux scheduler-specific wrapper structures `struct <sched_class>_rq`. For example, the CFS class defines a `struct cfs_rq` which is then declared inside of `struct rq` as `struct cfs_rq cfs`.
+
+The following snipt is taken from `/kernel/sched/sched.h`
+```
+/* CFS-related fields in a runqueue */
+struct cfs_rq {
+	struct load_weight	load;
+	unsigned long		runnable_weight;
+	unsigned int		nr_running;
+	unsigned int		h_nr_running;
+       /* code omited */
+};
+
+/*
+ * This is the main, per-CPU runqueue data structure.
+ *
+ * Locking rule: those places that want to lock multiple runqueues
+ * (such as the load balancing or the thread migration code), lock
+ * acquire operations must be ordered by ascending &runqueue.
+ */
```

Comment:
maybe for simplicity, get rid of the kernel comment before this struct definition. i dont want students thinking we wrote that and that they have to worry about locking. maybe just write this:
```
/*
 * This is the main, per-CPU runqueue data structure.
 *  ...rest of comments omitted 
 */
```